simpeg/optimization.py

```diff
@@ -1196,88 +1339,127 @@ def evalFunction(x, return_g=False):
-
-    @count
-    def activeSet(self, x):
-        """activeSet(x)
-
-        If we are on a bound
+        return self._step_active_set
 
-        """
-        return np.logical_or(x <= self.lower, x >= self.upper)
+    @step_active_set.setter
+    def step_active_set(self, value):
+        self._step_active_set = validate_type("step_active_set", value, bool)
 
     @property
-    def approxHinv(self):
-        """
-        The approximate Hessian inverse is used to precondition CG.
+    def active_set_grad_scale(self):
+        """Scalar to apply to the active set's gradient
 
-        Default uses BFGS, with an initial H0 of *bfgsH0*.
+        if `step_active_set` is `True`, then the active set's gradient is multiplied by this value
+        when including it in the search direction.
 
-        Must be a scipy.sparse.linalg.LinearOperator
+        Returns
+        -------
+        float
         """
-        _approxHinv = getattr(self, "_approxHinv", None)
-        if _approxHinv is None:
-            M = sp.linalg.LinearOperator(
-                (self.xc.size, self.xc.size), self.bfgs, dtype=self.xc.dtype
-            )
-            return M
-        return _approxHinv
+        return self._active_set_grad_scale
 
-    @approxHinv.setter
-    def approxHinv(self, value):
-        self._approxHinv = value
+    @active_set_grad_scale.setter
+    def active_set_grad_scale(self, value):
+        self._active_set_grad_scale = validate_float(
+            "active_set_grad_scale", value, min_val=0, inclusive_min=True
+        )
 
     @timeIt
     def findSearchDirection(self):
         """
         findSearchDirection()
         Finds the search direction based on projected CG
         """
+        # remember, "active" means cell with values equal to the limit
+        # "inactive" are cells with values inside the limits.
+
+        # The basic logic of this method is to do CG iterations only
+        # on the inactive set, then also add a scaled gradient for the
+        # active set, (if that gradient points away from the limits.)
+
         self.cg_count = 0
         Active = self.activeSet(self.xc)
-        temp = sum((np.ones_like(self.xc.size) - Active))
+        Inactive = 1 - Active
 
         step = np.zeros(self.g.size)
-        resid = -(1 - Active) * self.g
+        resid = Inactive * (-self.g)
 
-        r = resid - (1 - Active) * (self.H * step)
+        r = resid  # - Inactive * (self.H * step)#  step is zero
 
         p = self.approxHinv * r
 
         sold = np.dot(r, p)
 
         count = 0
+        r_norm0 = norm(r)
 
-        while np.all([np.linalg.norm(r) > self.tolCG, count < self.maxIterCG]):
+        atol = max(self.cg_rtol * norm(r_norm0), self.cg_atol)
+        if self.debug:
+            print(f"CG Target tolerance: {atol}")
+        r_norm = r_norm0
+        while r_norm > atol and count < self.cg_maxiter:
+            if self.debug:
+                print(f"CG Iteration: {count}, residual norm: {r_norm}")
```

Comment:
Previous tolerance was always an absolute tolerance, making the behavior of this minimizer very problem dependent... This updated check is exactly equivalent to `scipy.sparse.linalg.cg`.

Reply:
So can we just remove the CG while loop here? Feels like we could just multiple the `self.approxHinv` with the `inactive` and pass it to scipy?